src/hotspot/share/oops/constantPool.cpp

```diff
@@ -548,6 +548,11 @@ Klass* ConstantPool::klass_at_impl(const constantPoolHandle& this_cp, int which,
     }
   }
 
+  // We need to recheck exceptions from racing thread and return the same.
+  if (this_cp->tag_at(which).is_unresolved_klass_in_error()) {
+    throw_resolution_error(this_cp, which, CHECK_NULL);
+  }
+
```

Comment:
I'm unclear how this race is resolved. There must be a serialization point between the two threads otherwise this re-check is just as racy as the original. We need to know that the CP entry is now stable and was resolved either by this thread or the other one. But I can't see where this serialization point arises. ??

Reply:
Yes this is the wrong commit.  Sorry I thought it would help to have the separate commits.  The first commit should only add the test.